Shared/ActiveDirectoryFunctions/Test-ADCredentials.ps1

```diff
@@ -0,0 +1,89 @@
+
+    begin {
+        Write-Verbose "Calling: $($MyInvocation.MyCommand)"
+        $credentialsValid = $null
+        # Username formats: https://learn.microsoft.com/windows/win32/secauthn/user-name-formats
+        $usernameFormat = "local"
+        try {
+            Add-Type -AssemblyName System.DirectoryServices.Protocols -ErrorAction Stop
+        } catch {
+            Write-Verbose "Failed to load System.DirectoryServices.Protocols"
+            Invoke-CatchActionError $CatchActionFunction
+        }
+    }
+    process {
+        $domain = $Credentials.GetNetworkCredential().Domain
+        if ([System.String]::IsNullOrEmpty($domain)) {
+            Write-Verbose "Domain is empty which could be an indicator that UPN was passed instead of domain\username"
+            $domain = ($Credentials.GetNetworkCredential().UserName).Split("@")
+            if ($domain.Count -eq 2) {
+                Write-Verbose "Domain was extracted from UPN"
+                $domain = $domain[-1]
+                $usernameFormat = "upn"
+            } else {
+                Write-Verbose "Failed to extract domain from UPN - seems that username was passed without domain and so cannot be validated"
+                $domain = $null
+            }
+        } else {
+            Write-Verbose "Username was provided in down-level logon name format"
+            $usernameFormat = "downlevel"
+        }
+
+        if (-not([System.String]::IsNullOrEmpty($domain))) {
+            $ldapDirectoryIdentifier = New-Object System.DirectoryServices.Protocols.LdapDirectoryIdentifier($domain)
+            # Use Kerberos authentication as NTLM might lead to false/positive results in case the password was changed recently
+            $ldapConnection = New-Object -TypeName System.DirectoryServices.Protocols.LdapConnection($ldapDirectoryIdentifier, $Credentials, [DirectoryServices.Protocols.AuthType]::Kerberos)
+            # Enable Kerberos encryption (sign and seal)
+            $ldapConnection.SessionOptions.Signing = $true
+            $ldapConnection.SessionOptions.Sealing = $true
+            try {
+                $ldapConnection.Bind()
+                Write-Verbose "Connection succeeded with credentials"
+                $credentialsValid = $true
+            } catch [System.DirectoryServices.Protocols.LdapException] {
+                if ($_.Exception.ErrorCode -eq 49) {
+                    # ErrorCode 49 means invalid credentials
+                    Write-Verbose "Failed to connect to LDAP server with credentials provided"
+                    $credentialsValid = $false
+                } else {
+                    Write-Verbose "Failed to connect to LDAP server for other reason"
+                }
+                Invoke-CatchActionError $CatchActionFunction
+            } catch {
+                Write-Verbose "Exception occurred while connecting to LDAP server - unable to perform credential validation"
+                Invoke-CatchActionError $CatchActionFunction
```

Comment:
We should have an inner exception in Write-Verbose for the things we don't know here. Just incase the caller doesn't do anything with `$CatchActionFunction`, like what is currently happening.

Reply:
Done